gn2/wqflask/oauth2/checks.py

```diff
@@ -12,29 +12,41 @@
     authserver_uri,
     oauth2_clientid,
     oauth2_clientsecret)
+from .request_utils import authserver_authorise_uri
 
 
 def require_oauth2(func):
     """Decorator for ensuring user is logged in."""
     @wraps(func)
     def __token_valid__(*args, **kwargs):
         """Check that the user is logged in and their token is valid."""
-
         def __clear_session__(_no_token):
             session.clear_session_info()
             flash("You need to be logged in.", "alert-warning")
             return redirect("/")
 
+        def __redirect_to_login__(_token):
+            """
+            Save the current user request to session then
+            redirect to the login page.
+            """
+            if request.method == "GET":
+                redirect_url = url_for(request.endpoint, **request.args)
+            else:
+                redirect_url = "/"
```

Comment:
A minor improvement would be to assume all requests have a corresponding `GET`, so we could change this to:
```
# Note: this isn't tested but writen to demonstrate the idea
try:
  redirect_url = url_for(rquest.endpoint, _method="GET", **request.args)
except BuildError:
  redirect_url = "/"
```

Here's the tests I've done for this:
```
(Pdb) url_for("gsearch_updating", _method="GET")
*** werkzeug.routing.BuildError: Could not build url for endpoint 'gsearch_updating' ('GET'). Did you mean to use methods ['OPTIONS', 'POST']?
(Pdb) url_for("gsearch_updating", _method="POST")
'/gsearch_updating'
```

The advantage is that for endpoints that support both `GET` and `POST`, we'll redirect to the page that had the form

Reply:
Thanks, made the requested changes.